### PR TITLE
FIR checker: report COMPONENT_FUNCTION_RETURN_TYPE_MISMATCH

### PR DESCRIPTION
**Pull request by jsjeon**
When destructuring declarations, one can declare what type is expected (per component), or it could be inferred from other contexts, e.g., use-site. If the types mismatch, we report:
* COMPONENT_FUNCTION_RETURN_TYPE_MISMATCH

Unlike other `COMPONENT_CALL*` variants, which emerge as error named references, that diagnostic has a resolved component call. If that resolved component call returns incompatible type with an expected type, we should report that diagnostic.

### CODE REVIEW

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt [hunk 1]
```diff
@@ -75,39 +73,69 @@ object FirDestructuringDeclarationChecker : FirPropertyChecker() {
...
+        when (val reference = componentCall.calleeReference) {
+            is FirResolvedNamedReference -> {
+                val destructuringCall = (reference.resolvedSymbol.fir as? FirSimpleFunction) ?: return
+                val destructuringType = destructuringCall.returnTypeRef.coneType
+                if (destructuringType is ConeKotlinErrorType) {
+                    // There will be other errors on this error type.
+                    return
+                }
+                if (destructuringType is ConeTypeParameterType) {
+                    // As long as all involved constraints meet, this type parameter shouldn't be type mismatch,
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt:85]: Consider:
```kt
data class C<T>(val x: Int, val y: T)
fun <T, S> foo(c: C<T>) {
  val (x: Int, y: S) = c
}
```

FE1.0 has `COMPONENT_FUNCTION_RETURN_TYPE_MISMATCH` error on `c`, but I think this is the case inside this block (`destructuringType is ConeTypeParameterType`). So I think we still need to check the declaration and destructuring types.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt [hunk 2]
```diff
@@ -75,39 +73,69 @@ object FirDestructuringDeclarationChecker : FirPropertyChecker() {
...
-                        ),
+        when (val reference = componentCall.calleeReference) {
+            is FirResolvedNamedReference -> {
+                val destructuringCall = (reference.resolvedSymbol.fir as? FirSimpleFunction) ?: return
+                val destructuringType = destructuringCall.returnTypeRef.coneType
+                if (destructuringType is ConeKotlinErrorType) {
+                    // There will be other errors on this error type.
+                    return
+                }
+                if (destructuringType is ConeTypeParameterType) {
```
Comment (tgeng) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt:84]: This doesn't seem to work for the following code. `i:T` should be highlighted inside the call of `foo` in `bar`.

```
data class A<T>(val i: T)

fun <T> foo(block: (A<T>) -> Unit) {}

fun <T, R> bar() {
    foo<R> { (i: T) ->
    }
}
```

Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt:84]: Oops! Now this comment is redundant with https://github.com/JetBrains/kotlin/pull/4286#discussion_r608162295

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt]: Thanks. It turns out `componentCall` already has a substituted type, so we don't need special handling for type parameter. Also, it looks like there isn't a test like that, thanks for the example!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt (hunk 2 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt]: Yep, same concern. Added your example as well, thanks!

**Review by punzki — APPROVED** [review]

**Review by tgeng — APPROVED** [review]

**Review by mglukhikh — APPROVED** [review]
Looks correct, I'll run some extra tests and merge it.

**Comment by mglukhikh** [comment]
Merged manually. I had to add the following commit to fix test data: https://github.com/JetBrains/kotlin/commit/5229d4e4f4ad1a30c2727df594f73ac50b47ebf2. Could you please investigate what happened with LT diagnostic here?

**Comment by jsjeon** [comment]
> Merged manually. I had to add the following commit to fix test data: [5229d4e](https://github.com/JetBrains/kotlin/commit/5229d4e4f4ad1a30c2727df594f73ac50b47ebf2). Could you please investigate what happened with LT diagnostic here?

Will do, thanks.

**Comment by jsjeon** [comment]
> > Merged manually. I had to add the following commit to fix test data: [5229d4e](https://github.com/JetBrains/kotlin/commit/5229d4e4f4ad1a30c2727df594f73ac50b47ebf2). Could you please investigate what happened with LT diagnostic here?
> 
> Will do, thanks.

What a simple root cause... https://github.com/JetBrains/kotlin/pull/4297 I should have ran tests more thoroughly after seeing https://github.com/JetBrains/kotlin/pull/4288

**Pull request closed without merging**